File: 1351-CountNegativeNumbersInASortedMatrix/1351-CountNegativeNumbersInASortedMatrix.py
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        res = 0
        r = n - 1
        c = 0

        while r != -1:
            while c < m and grid[r][c] >= 0:
                c += 1
            else:
                res += m - c
                r -= 1

        return res

File: 1351-CountNegativeNumbersInASortedMatrix/1351-CountNegativeNumbersInASortedMatrix.exs
defmodule Solution do
  @spec count_negatives(grid :: [[integer]]) :: integer
  def count_negatives(grid) do
    {n, m} = {length(grid), length(hd(grid))}
    count(grid, n - 1, 0, 0, m)
  end
  defp count(_grid, -1, _c, res, _m) do res end
  defp count(grid, r, c, res, m) when c >= m do
    count(grid, r - 1, c, res, m)
  end
  defp count(grid, r, c, res, m) do
    row = Enum.at(grid, r)
    num = Enum.at(row, c)

    if num >= 0 do
      count(grid, r, c + 1, res, m)
    else
      count(grid, r - 1, c, res + m - c, m)
    end
  end
end
